Use Reactive Contexts instead of doing our own context propagation for rxjava

FroMage opened the pull request:
As discussed on mailing list:

At the moment, the resteasy-rxjava and resteasy-rxjava2 modules register hooks into the rxjava and rxjava2 plugin/hook system, in order to propagate the RESTEasy context (thread-local) into all phases of rxjava single/flowable/etc, which can otherwise be scheduled on any scheduler/thread and so would lose the RESTEasy context.

RxJava being what it is, you can only register a single plugin/hook globally, so if RESTEasy defines it, nobody else can. That's problematic, because CDI also requires context propagation, and so does Redpipe (to name just the two examples I am using ATM), so I created a library called Reactive Contexts which decouples libraries that have a context to propagate (RESTEasy, CDI, Redpipe, etc…) and libraries that provide context propagation (RxJava1, 2, etc…).

https://reactiverse.io/reactive-contexts

That library is super small (4k core, 13k propagator for rxjava). I'd like to remove the custom context propagation in the resteasy-rxjava/rxjava2 and make those modules depend on a new resteasy-reactive-context which would depend on reactive-contexts-core to provide a context provider for RESTEasy.

This way, I can also make CDI and Redpipe provide such a module and all those contexts will be propagated for rxjava for all users :)

WDYT? Do you agree on that extra dependency ? It's only for the rxjava modules, not the core.

marekkopecky commented on resteasy-reactive-context/src/test/java/org/resteasy/reactive/context/AppTest.java:
```diff
@@ -0,0 +1,38 @@
+     */
+    public void testApp()
+    {
+        assertTrue( true );
```
It would be better to remove this at all :)

Or even better, write resteasy context test to testsuite module.

Or even better, write resteasy && cdi context test to testsuite module.

FroMage replied:
Ah yes, that's a mistake (added by my IDE). Sorry, I'll remove it.

marekkopecky commented:
* Community documentation needs to be updated (maybe in some new PR in the future, but before release).
* What about 3.6 branch?
* @asoldano If you plan to use this in WF, new WF modules needs to be created: resteasy-reactive-context, reactive-contexts-core
* @FroMage When do you plan to release reactive-contexts 1.0? Currently, release reactive-contexts is 0.0.2. Is (will be) this project (reactive-contexts) backward compatible?

marekkopecky commented:
Link to RESTEASY jira is missing in PR description&&title&&commit_name

FroMage commented:
I removed the empty test class, and added the resteasy-reactive-context module, but I still have failing tests:

```
  NoStreamRxTest.testRx:65 » InternalServerError HTTP 500 Internal Server Error
  NoStreamRx2Test.testSingle:68 » InternalServerError HTTP 500 Internal Server E...
```

What's annoying is that they work if I run them alone, from the CLI and from the IDE. I'm pretty sure it's due to a services issue, because I had to add an `import services` to the manifest. @marekkopecky are you sure we have to deal with services that way in the war manifest? Shouldn't we declare services in the module descriptor? That sounds error-prone for every user, if they forget it, no?

kanovotn commented:
I like the idea of reactive-contexts project! With new library used by RESTEasy and possibly included in Wildfly there come questions related to the fact that the new library should be maintained long term:

- Who is going to merge PR's to the project?
- Who is going to make releases?
- Can somebody from resteasy developers having push rights?

The above are imho important to know before we start using it in resteasy, since rx java support will be  tied with it with this project and this is tied with support in Wildfly, in case of some critical issue we may need to act quickly.

FroMage commented:
At the moment, I'm merging PRs, making releases, and I guess I could be called a resteasy developer with push rights too :) But it's a reactiverse project, with an open community. If anyone else from the resteasy developer group wants to help maintain it, I'd be glad to let them.

rsvoboda commented on resteasy-rxjava2/pom.xml:
```diff
@@ -63,6 +63,15 @@
     </dependency>
+    <dependency>
+        <groupId>org.jboss.resteasy</groupId>
+        <artifactId>resteasy-reactive-context</artifactId>
```
Just cosmetic comment, this should probably go up to other `org.jboss.resteasy` dependencies

deps should be ordered alphabetically by GA(V)

rsvoboda commented on resteasy-dependencies-bom/pom.xml:
```diff
@@ -745,6 +746,21 @@
                 <artifactId>json-patch</artifactId>
                 <version>${version.json.patch}</version>
             </dependency>
+            <dependency>
```
These deps could probably stay here, there are more deps which are not sorted alphabetically. @asoldano wdyt?

FroMage commented:
@asoldano do you have any idea why my tests would fail if I run NoStreamRx2Test and NoStreamRxTest in a row? It very much looks like a jboss modules/classloader issue. I can run both independently and it works, but not both at the same time.

It is probably because the service loader initialisation of `Context` is done in a static block, and so probably it gets initialised for the first test with only the rxjava1 propagator being found, and when the second test runs with the rxjava2 propagator being present, the `Context` class isn't re-initialised so it doesn't find the rxjava2 propagator. At least that seems to be the likely cause, as of course this only happens in wildfly and not in a debugger.

But I thought classloading would mean each tested application with arquillian would get its own resteasy-reactive-context (and thus reactive-contexts-core) classloader?

FroMage commented:
I confirm it's related to jboss modules. There's something I just don't get. If I start a single test, the right rx hooks are called. If I start two, only one hook is called (rx1 or rx2 but not both). No matter if I start one or two tests though, it appears the `Context` class is initialised twice:

```
[0m[0m18:14:59,008 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-4) WFLYSRV0027: Starting deployment of "NoStreamRxTest.war" (runtime-name: "NoStreamRxTest.war")
[0m[33m18:14:59,383 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-4) WFLYSRV0018: Deployment "deployment.NoStreamRxTest.war" is using a private module ("org.jboss.resteasy.resteasy-json-binding-provider") which may be changed or removed in future versions without notice.
[0m[33m18:14:59,548 WARN  [org.jboss.weld.deployer] (MSC service thread 1-1) WFLYWELD0013: Deployment NoStreamRxTest.war contains CDI annotations but no bean archive was found (no beans.xml or class with bean defining annotations was present).
[0m[0m18:14:59,683 INFO  [org.infinispan.factories.GlobalComponentRegistry] (MSC service thread 1-6) ISPN000128: Infinispan version: Infinispan 'Bastille' 9.1.6.Final
[0m[0m18:14:59,803 INFO  [org.jboss.as.clustering.infinispan] (ServerService Thread Pool -- 69) WFLYCLINF0002: Started client-mappings cache from ejb container
[0m[31m18:15:00,202 ERROR [stderr] (ServerService Thread Pool -- 4) Loading context!!
[0m[31m18:15:00,203 ERROR [stderr] (ServerService Thread Pool -- 4) -> Provider: class org.resteasy.reactivecontext.ResteasyContextProvider
[0m[31m18:15:00,207 ERROR [stderr] (ServerService Thread Pool -- 4) -> Propagator: class io.reactiverse.reactivecontexts.propagators.rxjava1.RxJava1ContextPropagator
[0m[0m18:15:00,267 INFO  [org.jboss.resteasy.resteasy_jaxrs.i18n] (ServerService Thread Pool -- 4) RESTEASY002225: Deploying javax.ws.rs.core.Application: class org.jboss.resteasy.utils.TestApplication
[0m[0m18:15:00,270 INFO  [org.jboss.resteasy.resteasy_jaxrs.i18n] (ServerService Thread Pool -- 4) RESTEASY002200: Adding class resource org.jboss.resteasy.test.rx.rxjava.resource.RxNoStreamResource from Application class org.jboss.resteasy.utils.TestApplication
[0m[0m18:15:00,296 INFO  [org.hibernate.validator.internal.util.Version] (ServerService Thread Pool -- 4) HV000001: Hibernate Validator 5.3.6.Final
[0m[0m18:15:00,323 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 4) WFLYUT0021: Registered web context: '/NoStreamRxTest' for server 'default-server'
[0m[0m18:15:00,389 INFO  [org.jboss.as.server] (management-handler-thread - 1) WFLYSRV0010: Deployed "NoStreamRxTest.war" (runtime-name : "NoStreamRxTest.war")
[0mLoading context!!
-> Provider: class org.resteasy.reactivecontext.ResteasyContextProvider
-> Propagator: class io.reactiverse.reactivecontexts.propagators.rxjava1.RxJava1ContextPropagator
-> Propagator: class io.reactiverse.reactivecontexts.propagators.rxjava2.RxJava2ContextPropagator
```

This is with a single rx1 test, so no idea why it's initialised twice with different services too. Pretty sure the rx test declares only a single rxjava1 services, so no idea why the rx2 propagator even appears here.

FroMage commented:
Ah, I guess the second one is from the client code, running in JUnit. While the first one is from jboss modules in WF. Pity that the server logs get overwritten after each test deploy :(

rsvoboda commented:
Just a tip about the logs, https://github.com/resteasy/Resteasy/blob/master/pom.xml#L23
Try to take a look at target/surefire-reports/*-output.txt content

FroMage commented:
Thanks, I do look at them, but the server init logs are in `null-output.txt` which is shared by all tests, and so it's overwritten on each new test :(

marekkopecky commented:
The question is: who fix the future bugs in reactive-contexts project, if you (@FroMage ) won't have a time to do that? The answer could be "@asoldano", if he agrees :)

Reminder:
* @FroMage When do you plan to release reactive-contexts 1.0? Currently, reactive-contexts is 0.0.2.
* Is (will be) this project (reactive-contexts) backward compatible?

FroMage commented:
No, actually the startup of the first test is in `null-output.txt`, then the output of the first test AND the startup of the second test is in the `firsttest-output.txt`, then the output of the second test AND the shutdown of the server is in `secondtest-output.txt`. It sort of makes sense, but wasn't what I was expecting.

FroMage commented:
> The question is: who fix the future bugs in reactive-contexts project, if you (@FroMage ) won't have a time to do that? The answer could be "@asoldano", if he agrees :)

I guess that's a question for him, then :) I suppose @ronsigal could also volunteer. Or we could say that the Reactiverse community would.

> @FroMage When do you plan to release reactive-contexts 1.0? Currently, reactive-contexts is 0.0.2.

I can release it as soon as required.

> Is (will be) this project (reactive-contexts) backward compatible?

I can make it backward compatible.

FroMage commented:
OK I believe I have found a fix for the WF use-case, I'll update this Monday.

FroMage commented:
While looking for an error, I came across `SseEventSourceImpl.setAlwasyReconnect` (note the typo) @ronsigal you want an issue for that?
Also, related to that, I had a `Feature` implementation that depended on having a non-null `ResteasyProviderFactory.getContextData(Dispatcher.class)`, which is always the case for resteasy server, BUT is not the case for `SseEventSourceImpl` which runs things in a custom thread with no thread-local context:

```
java.lang.RuntimeException: java.lang.NullPointerException
	at org.jboss.resteasy.plugins.providers.RegisterBuiltin.register(RegisterBuiltin.java:49)
	at org.jboss.resteasy.spi.ResteasyProviderFactory.getInstance(ResteasyProviderFactory.java:741)
	at org.jboss.resteasy.plugins.providers.sse.InboundSseEventImpl.readData(InboundSseEventImpl.java:194)
	at org.jboss.resteasy.plugins.providers.sse.InboundSseEventImpl.readData(InboundSseEventImpl.java:182)
	at org.jboss.resteasy.plugins.providers.sse.InboundSseEventImpl.readData(InboundSseEventImpl.java:172)
	at org.jboss.resteasy.plugins.providers.sse.InboundSseEventImpl.toString(InboundSseEventImpl.java:236)
	at java.lang.String.valueOf(String.java:2994)
	at java.lang.StringBuilder.append(StringBuilder.java:131)
	at org.jboss.resteasy.plugins.providers.sse.client.SseEventSourceImpl$EventHandler.run(SseEventSourceImpl.java:398)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at org.resteasy.reactivecontext.ContextFeature.configure(ContextFeature.java:25)
	at org.jboss.resteasy.spi.ResteasyProviderFactory.processProviderContracts(ResteasyProviderFactory.java:2015)
	at org.jboss.resteasy.spi.ResteasyProviderFactory.registerProvider(ResteasyProviderFactory.java:1711)
	at org.jboss.resteasy.spi.ResteasyProviderFactory.registerProvider(ResteasyProviderFactory.java:1661)
	at org.jboss.resteasy.plugins.providers.RegisterBuiltin.registerProviders(RegisterBuiltin.java:133)
	at org.jboss.resteasy.plugins.providers.RegisterBuiltin.register(RegisterBuiltin.java:45)
	... 15 more
```

So this NPE got thrown because without the ResteasyProviderFactory set in the thread-local, it uses `getInstance()` for a singleton, which doesn't have a `Dispatcher`. So I had an NPE in my `Feature` and due to the `Executor` used in `SseEventSourceImpl` the exception was never caught, nor reported: simply ignored and it cost me three days to find. @ronsigal again, want an issue for that?

FroMage commented:
@asoldano uh, not sure what happened, but it looks like rxjava1 module is gone from jboss-modules. how is that supposed to work under wildfly then?

asoldano commented:
@FroMage wildfly is going to have the rxjava2 module only. If users want the rxjava1, they are expected to include the needed libraries in their deployments. The changes that applied this are at https://github.com/resteasy/Resteasy/commit/7b006cfe4856bc13c69c225a0297c4641b8a7132 ; in particular, at https://github.com/resteasy/Resteasy/commit/7b006cfe4856bc13c69c225a0297c4641b8a7132#diff-b2954a3f1f2c54bcc792a24ec605a398R14 you see the code used by the testsuite to setup deployments needing rxjava1 properly.

FroMage commented:
Hi, just rebased on the latest master, and fixed all tests.

asoldano commented on resteasy-reactive-context/src/main/java/org/resteasy/reactivecontext/ContextFeature.java:
```diff
@@ -0,0 +1,32 @@
+         return false;
+      Dispatcher dispatcher = ResteasyProviderFactory.getContextData(Dispatcher.class);
+      if(dispatcher == null) {
+         System.err.println("Creating new context for "+context.getConfiguration().getRuntimeType()+" deployment because it looks like we're in a special thread");
```
This should probably be i18n log

asoldano commented on jboss-modules/build.xml:
```diff
@@ -146,10 +146,25 @@
+        	<maven-resource group="io.reactivex" artifact="rxjava-reactive-streams"/>
+            <maven-resource group="io.reactiverse" artifact="reactive-contexts-propagators-rxjava1"/>
+        </module-def>
+        -->
```
maybe remove the commented out rxjava module stuff here

FroMage commented:
Opened https://issues.jboss.org/projects/RESTEASY/issues/RESTEASY-1879 for this.

FroMage commented:
Opened https://issues.jboss.org/browse/RESTEASY-1940 and https://issues.jboss.org/browse/RESTEASY-1941 for the other noted issues.

rsvoboda commented:
> Opened https://issues.jboss.org/projects/RESTEASY/issues/RESTEASY-1879 for this.

Did you mean https://issues.jboss.org/browse/RESTEASY-1939 ?

FroMage commented:
> Did you mean https://issues.jboss.org/browse/RESTEASY-1939 ?

Ah yes, absolutely. Sorry, JIRA must have tripped me with its weird UI.

FroMage commented:
Thanks!

asoldano commented:
@FroMage thank you; still need to think whether it's better to merge this to 3.6 or not, will figure out this week.

FroMage commented:
As you wish: I have no personal need for it to be in 3.6.

ronsigal commented:
Hi Stéphane,

I'm cleaning up my mailbox and found this. Is it still alive?

-Ron

On 6/25/18 10:48 AM, Stéphane Épardaud wrote:
>
> While looking for an error, I came across 
> |SseEventSourceImpl.setAlwasyReconnect| (note the typo) @ronsigal 
> <https://github.com/ronsigal> you want an issue for that?
> Also, related to that, I had a |Feature| implementation that depended 
> on having a non-null 
> |ResteasyProviderFactory.getContextData(Dispatcher.class)|, which is 
> always the case for resteasy server, BUT is not the case for 
> |SseEventSourceImpl| which runs things in a custom thread with no 
> thread-local context:
>
> |java.lang.RuntimeException: java.lang.NullPointerException at 
> org.jboss.resteasy.plugins.providers.RegisterBuiltin.register(RegisterBuiltin.java:49) 
> at 
> org.jboss.resteasy.spi.ResteasyProviderFactory.getInstance(ResteasyProviderFactory.java:741) 
> at 
> org.jboss.resteasy.plugins.providers.sse.InboundSseEventImpl.readData(InboundSseEventImpl.java:194) 
> at 
> org.jboss.resteasy.plugins.providers.sse.InboundSseEventImpl.readData(InboundSseEventImpl.java:182) 
> at 
> org.jboss.resteasy.plugins.providers.sse.InboundSseEventImpl.readData(InboundSseEventImpl.java:172) 
> at 
> org.jboss.resteasy.plugins.providers.sse.InboundSseEventImpl.toString(InboundSseEventImpl.java:236) 
> at java.lang.String.valueOf(String.java:2994) at 
> java.lang.StringBuilder.append(StringBuilder.java:131) at 
> org.jboss.resteasy.plugins.providers.sse.client.SseEventSourceImpl$EventHandler.run(SseEventSourceImpl.java:398) 
> at 
> java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) 
> at java.util.concurrent.FutureTask.run(FutureTask.java:266) at 
> java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) 
> at 
> java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) 
> at 
> java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) 
> at 
> java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) 
> at java.lang.Thread.run(Thread.java:748) Caused by: 
> java.lang.NullPointerException at 
> org.resteasy.reactivecontext.ContextFeature.configure(ContextFeature.java:25) 
> at 
> org.jboss.resteasy.spi.ResteasyProviderFactory.processProviderContracts(ResteasyProviderFactory.java:2015) 
> at 
> org.jboss.resteasy.spi.ResteasyProviderFactory.registerProvider(ResteasyProviderFactory.java:1711) 
> at 
> org.jboss.resteasy.spi.ResteasyProviderFactory.registerProvider(ResteasyProviderFactory.java:1661) 
> at 
> org.jboss.resteasy.plugins.providers.RegisterBuiltin.registerProviders(RegisterBuiltin.java:133) 
> at 
> org.jboss.resteasy.plugins.providers.RegisterBuiltin.register(RegisterBuiltin.java:45) 
> ... 15 more |
>
> So this NPE got thrown because without the ResteasyProviderFactory set 
> in the thread-local, it uses |getInstance()| for a singleton, which 
> doesn't have a |Dispatcher|. So I had an NPE in my |Feature| and due 
> to the |Executor| used in |SseEventSourceImpl| the exception was never 
> caught, nor reported: simply ignored and it cost me three days to 
> find. @ronsigal <https://github.com/ronsigal> again, want an issue for 
> that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/resteasy/Resteasy/pull/1572#issuecomment-399978504>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAg38zhfii0XRuJfp-ebllX_UszaIaceks5uAPgqgaJpZM4Ur1Nw>.
>
-- 
My company's smarter than your company (unless you work for Red Hat)
